governance/contract_connector.go

```
@@ -0,0 +1,125 @@
+package governance
+
+import (
+	"errors"
+	"math/big"
+	"strings"
+
+	"github.com/klaytn/klaytn/accounts/abi"
+	"github.com/klaytn/klaytn/blockchain"
+	"github.com/klaytn/klaytn/blockchain/types"
+	"github.com/klaytn/klaytn/blockchain/vm"
+	"github.com/klaytn/klaytn/common"
+	govcontract "github.com/klaytn/klaytn/contracts/gov"
+	"github.com/klaytn/klaytn/params"
+)
+
+type contractCaller struct {
+	chainConfig  *params.ChainConfig
+	chain        blockChain
+	contractAddr common.Address
+}
+
+var govParamAbi, _ = abi.JSON(strings.NewReader(govcontract.GovParamABI))
+
+func (c *contractCaller) getAllParamsAt(num *big.Int) (*params.GovParamSet, error) {
+	tx, err := c.makeTx(govParamAbi, "getAllParamsAt", num)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := c.callTx(tx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse results into GovParamSet
+
+	// Cannot parse empty data
+	if len(res) == 0 {
+		return params.NewGovParamSet(), nil
+	}
+
+	var ( // c.f. contracts/gov/GovParam.go:GetAllParams()
+		pNames  = new([]string)                   // *[]string = nil
+		pValues = new([][]byte)                   // *[][]byte = nil
+		out     = &[]interface{}{pNames, pValues} // array of pointers
+	)
+	if err := govParamAbi.Unpack(out, "getAllParams", res); err != nil {
+		return nil, err
+	}
+	var ( // Retrieve the slices allocated inside Unpack().
+		names  = *pNames
+		values = *pValues
+	)
+
+	if len(names) != len(values) {
+		logger.Error("Malformed contract.getAllParams result",
+			"len(names)", len(names), "len(values)", len(values))
+		return params.NewGovParamSet(), nil
```

Maybe we should return an error here?

---

done